modules/core/src/core/transform.js

```diff
@@ -221,4 +304,118 @@ export default class Transform {
       });
     }
   }
+
+  // create buffer to access source texture's individual pixels.
+  _createElementIDBuffer(elementCount) {
+    if (!this.hasSourceTextures) {
+      return;
+    }
+    if (this.elementIDBuffer) {
+      this.elementIDBuffer.delete();
+    }
+    // TODO: can attribute be a uint
+    const elementIds = new Float32Array(elementCount);
+    elementIds.forEach((_, index, array) => {
+      array[index] = index;
+    });
+    this.elementIDBuffer = new Buffer(this.gl, {data: elementIds, size: 1});
+  }
+
+  // build and return shader releated parameters
+  _getShaders(props = {}) {
+    const {vs, uniforms} = this._processVertexShader(props.vs);
+    const fs = getShaderVersion(vs) === 300 ? FS300 : FS100;
+    const modules = this.hasSourceTextures ?
+      [transform].concat(props.modules || []) : props.modules;
+    return {vs, fs, modules, uniforms};
+  }
+
+  // scan and update vertex shader for texture atrributes.
+  _processVertexShader(vs) {
```

Comment:
This should probably go into shadertools. Recommend making this into a utility in a separate file for now.

Reply:
Lot of this processing is very specific to `Transform` usecase, so may be not useful for other modules. But still decouple with Transform class and moved it to separate file `transform-shader-utils.js`